File: src/main/java/algo/slidingWindow/LongestRepeatingCharacterReplacement_424.java
package algo.slidingWindow;

public class LongestRepeatingCharacterReplacement_424 {
  public static int characterReplacement(String s, int k) {
    var freq = new int[26];

    int max = 0;
    int i = 0, j = 0;
    while (i <= j && j < s.length()) {
      var c = s.charAt(j);
      freq[c - 'A']++;

      // (window size - max freq) is the number of characters we are going to replace
      int maxFreq = 0;
      for (var n : freq) {
        maxFreq = Math.max(maxFreq, n);
      }
      int windowSize = j - i + 1;
      int replacements = windowSize - maxFreq;
      if (replacements <= k) {
        max = Math.max(max, windowSize);
      } else {
        freq[s.charAt(i) - 'A']--;
        i++;
      }
      j++;
    }

    return max;
  }

  public static void main(String[] args) {
    System.out.println(characterReplacement("ABAB", 2));
    System.out.println(characterReplacement("AABABBA", 1));
    System.out.println(characterReplacement("AABABBA", 2));
    System.out.println(characterReplacement("ABAA", 0));
    System.out.println(characterReplacement("BAAA", 0));
  }
}

File: src/main/kotlin/algo/slidingWindow/LongestRepeatingCharacterReplacement_424.kt
package algo.slidingWindow

class KtLongestRepeatingCharacterReplacement_424 {
    fun characterReplacement(s: String, k: Int): Int {
        var maxLen = 0
        var left = 0; var right = 0
        val freq: MutableMap<Char, Int> = mutableMapOf()
        while (right < s.length) {
            freq[s[right]] = freq.getOrDefault(s[right], 0) + 1
            val window = right - left + 1
            val maxFreqChar = freq.maxBy { it.value }.key
            if (window - freq[maxFreqChar]!! <= k) {
                maxLen = maxLen.coerceAtLeast(window)
            } else {
                val ch = s[left]
                freq[ch] = freq[ch]!! - 1
                left++
            }
            right++
        }
        return maxLen
    }
}

fun main() {
    val replace = KtLongestRepeatingCharacterReplacement_424()
    println(replace.characterReplacement("ABAB", 2))
    println(replace.characterReplacement("AABABBA", 1))
}